src/yunohost/app.py

```diff
@@ -483,28 +483,30 @@ def app_upgrade(app=[], url=None, file=None, force=False):
-            if version.parse(app_current_version) >= version.parse(app_new_version) and not force:
-                logger.success(m18n.n('app_already_up_to_date', app=app_instance_name))
-                # Save update time
-                now = int(time.time())
-                app_setting(app_instance_name, 'update_time', now)
-                app_setting(app_instance_name, 'current_revision', manifest.get('remote', {}).get('revision', "?"))
-                continue
-            elif version.parse(app_current_version) > version.parse(app_new_version):
-                upgrade_type = "DOWNGRADE_FORCED"
-            elif app_current_version == app_new_version:
-                upgrade_type = "UPGRADE_FORCED"
-            elif "~ynh" in app_current_version and "~ynh" in app_new_version:
-                app_current_version_upstream, app_current_version_pkg = app_current_version.split("~ynh")
-                app_new_version_upstream, app_new_version_pkg = app_new_version.split("~ynh")
-                if app_current_version_upstream == app_new_version_upstream:
-                    upgrade_type = "UPGRADE_PACKAGE"
-                elif app_current_version_pkg == app_new_version_pkg:
-                    upgrade_type = "UPGRADE_APP"
+            if "~ynh" in app_current_version and "~ynh" in app_new_version:
+                if version.parse(app_current_version) >= version.parse(app_new_version) and not force:
+                    # No new version available
+                    logger.success(m18n.n('app_already_up_to_date', app=app_instance_name))
+                    # Save update time
+                    now = int(time.time())
+                    app_setting(app_instance_name, 'update_time', now)
+                    app_setting(app_instance_name, 'current_revision', manifest.get('remote', {}).get('revision', "?"))
+                    continue
+                elif version.parse(app_current_version) > version.parse(app_new_version):
+                    upgrade_type = "DOWNGRADE_FORCED"
+                elif app_current_version == app_new_version:
+                    upgrade_type = "UPGRADE_FORCED"
                 else:
-                    upgrade_type = "UPGRADE_FULL"
+                    app_current_version_upstream, app_current_version_pkg = app_current_version.split("~ynh")
+                    app_new_version_upstream, app_new_version_pkg = app_new_version.split("~ynh")
+                    if app_current_version_upstream == app_new_version_upstream:
+                        upgrade_type = "UPGRADE_PACKAGE"
+                    elif app_current_version_pkg == app_new_version_pkg:
+                        upgrade_type = "UPGRADE_APP"
+                    else:
+                        upgrade_type = "UPGRADE_FULL"
+            else:
+                logger.warning("/!\\ Packagers ! You have enabled the setting 'upgrade_only_if_version_changes' but you haven't used the official way to define the package version")
```

Comment:
Uh wokay but it's not clear to me where that hard constrains comes from ? Granted that it's better to follow the convention, but `version.parse` is able to parse correctly a version with and without ~ynh...

After testing it, I do noticed that it will return some weird stuff for example here : 

```
>>> version.parse("1.2~ynh3") < version.parse("1.4")
True
>>> version.parse("1.2") < version.parse("1.4~ynh1")
False
```

(the second one ain't right...). Might be related to the fact that it parses it as "Legacy version" ?

But anyway, couldn't we hack it and concatenate `~ynh0` if there's no `~ynh` ? (And maybe display a warning if so)

Reply:
> (the second one ain't right...). Might be related to the fact that it parses it as "Legacy version" ?
> 
> But anyway, couldn't we hack it and concatenate `~ynh0` if there's no `~ynh` ? (And maybe display a warning if so)

My point of view is that, the packager need to enable explicitly this feature so for me in all case it's won't be enabled for old legacy package. So for me I don't think that it's really a problem to put this restriction. I think trying to manage some legacy package is not really a good idea as it might make issue than it solve.

Reply:
Meh okay :/ I can anticipate that some people will just yoloenable the flag "upgrade_only_if_version_changes" without paying attention to their version format because it's not intuitive that this format is required to support it...

https://paste.yunohost.org/raw/uzuvuqojol

But I'll at least add a check in package_linter about it ...

Reply:
> Meh okay :/ I can anticipate that some people will just yoloenable the flag "upgrade_only_if_version_changes" without paying attention to their version format because it's not intuitive that this format is required to support it...

It's why I added a warning.

The other problem is that (except for the `-ynh` instead of `~ynh`), don't think that it's really a good idea to enable this if there are not a package version in the version. If we do this we will go in a state when the app is not upgraded when the upstream version is not changed even if there are some change in the package version.

Reply:
Alrighty let's roll with this ! I added a check in packager_linter (and in the same process realizing that so many apps don't even have "version" defined ... :/ )